rpc/subscriptions_test.go

```diff
@@ -326,7 +330,533 @@ func TestSubscribeEvents(t *testing.T) {
+
+	conn1, _, err := websocket.Dial(ctx, httpSrv.URL, nil) //nolint:bodyclose
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		require.NoError(t, conn1.Close(websocket.StatusNormalClosure, ""))
+	})
+
+	conn2, _, err := websocket.Dial(ctx, httpSrv.URL, nil) //nolint:bodyclose
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		require.NoError(t, conn2.Close(websocket.StatusNormalClosure, ""))
+	})
+
+	firstID := uint64(1)
+	secondID := uint64(2)
+
+	handler.WithIDGen(func() uint64 { return firstID })
+	firstGot := sendWsMessage(t, ctx, conn1, subMsg("starknet_subscribeNewHeads"))
+	require.NoError(t, err)
+	require.Equal(t, subResp(firstID), firstGot)
+
+	handler.WithIDGen(func() uint64 { return secondID })
+	secondGot := sendWsMessage(t, ctx, conn2, subMsg("starknet_subscribeNewHeads"))
+	require.NoError(t, err)
+	require.Equal(t, subResp(secondID), secondGot)
+
+	// Ignore the first mock header
+	_, _, err = conn1.Read(ctx)
+	require.NoError(t, err)
+	_, _, err = conn2.Read(ctx)
+	require.NoError(t, err)
+
+	// Simulate a new block
+	syncer.newHeads.Send(testHeader(t))
+
+	// Receive a block header.
+	_, firstHeaderGot, err := conn1.Read(ctx)
+	require.NoError(t, err)
+	require.Equal(t, newHeadsResponse(firstID), string(firstHeaderGot))
+
+	_, secondHeaderGot, err := conn2.Read(ctx)
+	require.NoError(t, err)
+	require.Equal(t, newHeadsResponse(secondID), string(secondHeaderGot))
+
+	// Unsubscribe
+	unsubMsg := `{"jsonrpc":"2.0","id":1,"method":"juno_unsubscribe","params":[%d]}`
+	require.NoError(t, conn1.Write(ctx, websocket.MessageBinary, []byte(fmt.Sprintf(unsubMsg, firstID))))
+	require.NoError(t, conn2.Write(ctx, websocket.MessageBinary, []byte(fmt.Sprintf(unsubMsg, secondID))))
+}
+
+func TestSubscriptionReorg(t *testing.T) {
+	ctx, cancel := context.WithCancel(context.Background())
+	t.Cleanup(cancel)
+
+	mockCtrl := gomock.NewController(t)
+	t.Cleanup(mockCtrl.Finish)
+
+	mockChain := mocks.NewMockReader(mockCtrl)
+	syncer := newFakeSyncer()
+	handler, server := setupRPC(t, ctx, mockChain, syncer)
+
+	testCases := []struct {
+		name            string
+		subscribeMethod string
+		ignoreFirst     bool
+	}{
+		{
+			name:            "reorg event in starknet_subscribeNewHeads",
+			subscribeMethod: "starknet_subscribeNewHeads",
+			ignoreFirst:     true,
+		},
+		{
+			name:            "reorg event in starknet_subscribeEvents",
+			subscribeMethod: "starknet_subscribeEvents",
+			ignoreFirst:     false,
+		},
+		// TODO: test reorg event in TransactionStatus
+	}
+
+	mockChain.EXPECT().HeadsHeader().Return(&core.Header{}, nil).Times(len(testCases))
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			conn := createWsConn(t, ctx, server)
+
+			id := uint64(1)
+			handler.WithIDGen(func() uint64 { return id })
+
+			got := sendWsMessage(t, ctx, conn, subMsg(tc.subscribeMethod))
+			require.Equal(t, subResp(id), got)
+
+			if tc.ignoreFirst {
+				_, _, err := conn.Read(ctx)
+				require.NoError(t, err)
```

Comment:
Do you think just checking for no error is meaningful enough?

Reply:
Yes because I just want to test the header sent through the feed, so the first response received can be ignored.